core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1969,34 +1951,55 @@ class KafkaApis(val requestChannel: RequestChannel,
       throw new ClusterAuthorizationException(s"Request $request is not authorized.")
   }
 
-  private def sendResponseMaybeThrottle(request: RequestChannel.Request, createResponse: Int => AbstractResponse) {
-    sendResponseMaybeThrottle(request, request.header.clientId, { requestThrottleMs =>
-      sendResponse(request, createResponse(requestThrottleMs))
-    })
+  private def sendResponseMaybeThrottle(request: RequestChannel.Request, createResponse: Int => AbstractResponse): Unit = {
+    quotas.request.maybeRecordAndThrottle(request,
+      throttleTimeMs => sendResponse(request, Some(createResponse(throttleTimeMs))))
   }
 
-  private def sendResponseMaybeThrottle(request: RequestChannel.Request, clientId: String, sendResponseCallback: Int => Unit) {
-
-    if (request.apiRemoteCompleteTimeNanos == -1) {
-      // When this callback is triggered, the remote API call has completed
-      request.apiRemoteCompleteTimeNanos = time.nanoseconds
+  private def sendErrorResponseMaybeThrottle(request: RequestChannel.Request, error: Throwable) {
+    def sendError(throttleTimeMs: Int): Unit = {
+      val response = request.body[AbstractRequest].getErrorResponse(throttleTimeMs, error)
+      if (response == null)
+        closeConnection(request)
+      else
+        sendResponse(request, Some(response))
     }
-    quotas.request.maybeRecordAndThrottle(request.session.sanitizedUser, clientId,
-        request.requestThreadTimeNanos, sendResponseCallback,
-        callback => request.recordNetworkThreadTimeCallback = Some(callback))
+    quotas.request.maybeRecordAndThrottle(request, sendError)
+  }
+
+  private def sendResponseExemptThrottle(request: RequestChannel.Request, response: AbstractResponse): Unit = {
+    quotas.request.maybeRecordExempt(request)
+    sendResponse(request, Some(response))
   }
 
-  private def sendResponseExemptThrottle(response: RequestChannel.Response) {
-    sendResponseExemptThrottle(response.request, () => requestChannel.sendResponse(response))
+  private def sendErrorResponseExemptThrottle(request: RequestChannel.Request, error: Throwable): Unit = {
+    quotas.request.maybeRecordExempt(request)
+    val response = request.body[AbstractRequest].getErrorResponse(0, error)
+    if (response == null)
+      closeConnection(request)
+    else
+      sendResponse(request, Some(response))
```

Comment:
Seems like the 5 lines above are the same as the `sendError` method in `sendErrorResponseMaybeThrottle` so maybe we can extract a helper that takes `throttleTimeMs`.